index.bs

```diff
@@ -4109,165 +4310,1257 @@ To <dfn>get the response data</dfn> given |response|:
+
+        1. If |codepoint| is U+005B ([), set |inside brackets| to true.
+
+        1. If |codepoint| is U+005D (]), set |inside brackets| to false.
+
+      1. Append |hostname| to |pattern url|.
+
+    1. Otherwise:
+
+      1. If |scheme| is not "<code>file</code>", append "<code>placeholder</code>" to
+         |pattern url|.
+
+      1. Set |has hostname| to false.
+
+    1. If |pattern| [=map/contains=] "<code>port</code>":
+
+      1. If |pattern|["<code>port</code>"] is the empty string, return [=error=]
+         with [=error code=] [=invalid argument=].
+
+      1. Let |port| be the result of [=trying=] to [=unescape
+         URL Pattern=] with |pattern|["<code>port</code>"].
+
+      1. Append "<code>:</code>" to |pattern url|.
+
+      1. For each |codepoint| in |port|:
+
+        1. If |codepoint| is not an [=ASCII digit=]:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |port| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>pathname</code>":
+
+      1. Let |pathname| be the result of [=trying=] to [=unescape URL Pattern=] with
+         |pattern|["<code>pathname</code>"].
+
+      1. If |pathname| does not [=string/starts with|start with=] U+002F (/), then
+         append "<code>/</code>" to pattern.
+
+      1. For each |codepoint| in |pathname|:
+
+        1. If «U+003F (?), U+0023 (#)» [=set/contains=]
+           |codepoint|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |pathname| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>search</code>":
+
+      1. Let |search| be the result of [=trying=] to [=unescape URL pattern=] with
+         |pattern|["<code>search</code>"].
+
+      1. If |search| does not [=string/starts with|start with=] U+003F (?), then
+         append "<code>?</code>" to |pattern url|.
+
+      1. For each |codepoint| in |search|:
+
+        1. If |codepoint| is U+0023 (#):
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |search| to |pattern url|.
+
+1. Otherwise, if |pattern| matches the <code>network.UrlPatternString</code> production:
+
+  1. Let |pattern url| be the result of [=trying=] to [=unescape URL
+     pattern=] with |pattern|["<code>pattern</code>"].
+
+1. Let |url| be the result of [=URL Parser|parsing=] |pattern url|.
+
+1. If |url| is failure, return [=error=] with [=error code=] [=invalid argument=].
+
+1. Let |parsed| be a struct with the following fields:
+   <dl>
+     <dt>protocol
+     <dd>|url|'s [=url/scheme=] if |has protocol| is true, or null otherwise.
+     <dt>hostname
+     <dd>|url|'s [=url/host=] if |has hostname| is true, or null otherwise.
+     <dt>port
+     <dd>
+     1. If |url|'s [=url/scheme=] [=is special=] and |url|'s
+   [=url/scheme=]'s [=default port=] is not null, and |url|'s [=url/port=] is
+   null or is equal to [=url/scheme=]'s [=default port=]:
```

Comment:
I still feel like if no port is provided, we should match any port (ie have some has port logic). We do that for hostname and protocol, so maybe it would make sense here as well?

On top of that, we can have some odd scenarios when users did not provide a protocol. Since we default to `http` in that case, the algorithm will behave unexpectedly in some cases (eg if provided port is 80). Maybe we should also take `has protocol` into account in this condition?

Reply:
Should we maybe default to `https` those days?

Reply:
> Should we maybe default to `https` those days?

Note that In theory, `http` here really only acts as a placeholder so that we can build a URL that we can parse later. You will still have the `has protocol` flag to `false`, which means the pattern will not attempt to match on the protocol. However since this is a special scheme, it does have some side effects: 
- has a default port
- path defaulting to "/"

If we check `has protocol` in the condition as suggested above, then using http or https should be identical. And if not, https just moves the edge case to scenarios where users provide a URLPatterPattern such as `{ port: 443 }` which would start matching `http` or `ws` requests on port 80, `ftp` requests on port 21, which is just as unexpected.